Mathlib/Data/List/Shortlex.lean

```diff
@@ -0,0 +1,294 @@
+/-
+Copyright (c) 2024 Hannah Fechtner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hannah Fechtner
+-/
+
+import Mathlib.Data.List.Lex
+import Mathlib.Tactic.Linarith
+
+/-!
+# Shortlex ordering of lists.
+
+Given a relation `r` on `α`, the shortlex order on `List α` is defined by `L < M` iff
+* `L.length < M.length`
+* `L.length = M.length` and `L < M` under the lexicographic ordering over `r` on lists
+
+## Main results
+
+We show that if `r` is well-founded, so too is the shortlex order over `r`
+
+## See also
+
+Related files are:
+* `Mathlib.Data.List.Lex`: Lexicographic order on `List α`.
+* `Mathlib.Data.DFinsupp.WellFounded`: Well-foundedness of lexicographic orders on `DFinsupp` and
+  `Pi`.
+-/
+
+/-! ### shortlex ordering -/
+
+namespace List
+
+/-- Given a relation `r` on `α`, the shortlex order on `List α`, for which
+`[a0, ..., an] < [b0, ..., b_k]` if `n < k` or `n = k` and `[a0, ..., an] < [b0, ..., bk]`
+under the lexicographic order induced by `r`. -/
+def Shortlex {α : Type*} (r : α → α → Prop) : List α → List α → Prop :=
+  fun a b => Prod.Lex (fun n1 n2 => n1 < n2) (fun a b => List.Lex r a b) (a.length, a) (b.length, b)
+
+namespace Shortlex
+
+variable {α : Type*} {r : α → α → Prop}
+
+/-- If a list `s` is shorter than a list `t`, then `s` is smaller than `t` under any shortlex
+order -/
+theorem shorter_than {s t : List α} (h : s.length < t.length) : Shortlex r s t :=
+  Prod.Lex.left _ _ h
+
+/-- If two lists `s` and `t` have the same length, `s` is smaller than `t` under the shortlex order
+over a relation `r` exactly when `s` is smaller than `t` under the lexicographic order over `r`-/
+theorem len_equal {s t : List α} (h : s.length = t.length) (h2 : List.Lex r s t) :
+    Shortlex r s t := by
+  unfold Shortlex
+  rw [h]
+  exact Prod.Lex.right _ h2
+
+theorem shortlex_def {s t : List α} : Shortlex r s t ↔
+    s.length < t.length ∨ s.length = t.length ∧ List.Lex r s t := by
+  constructor
+  · intro hs
+    unfold Shortlex at hs
+    generalize hp : (s.length, s) = p at hs
+    cases hs with
+    | left b₁ b₂ h =>
+      left
+      rw [Prod.mk.injEq] at hp
+      rw [← hp.1] at h
+      exact h
+    | right a h =>
+      right
+      rw [Prod.mk.injEq] at hp
+      rw [← hp.2] at h
+      exact ⟨hp.1, h⟩
+  intro hpq
+  rcases hpq with h1 | h2
+  · exact shorter_than h1
+  exact len_equal h2.1 h2.2
+
+theorem cons_iff [IsIrrefl α r] {a : α} {s t : List α} : Shortlex r (a :: s) (a :: t) ↔
+    Shortlex r s t := by
+  constructor
```

Comment:
I would guess you can make lots of progress here with `simp [shortlex_def]`

Reply:
`simp [shortlex_def, List.Lex.cons_iff]` does the whole thing, wow!